### PR TITLE
Introduce wicketd

### PR DESCRIPTION
**Pull request by andrewjstone**
`wicketd` is the backend service that wicket will interact with to provide technician port based operator functionality to the rack. This is the start of an implementation and there is much more to do. Currently, `wicketd` is essentially a caching MGS proxy for the `/sp` endpoint of MGS. The MGS endpoint is paginated, but for wicket, we don't really mind waiting as we plan to poll for changes periodically. To ensure we get all information, we just attempt to retrieve the max amount of components (36) within a given timeout. This limits us to one page that wicket then stores in memory and serves as `/inventory`.

There is an open TODO to fetch the SP inventory for each SP retrieved from `/sp` via the MGS `/sp/{type}/{slot}/component` endpoint. Once this is implemented, wicket will be modified to talk to wicketd and serve this inventory instead of the fake inventory currently there.

Since we want to allow proxying information from MGS with a minimum of hassle, `gateway-client` was changed to derive `schemars::JsonSchema` for generated types so those could be served from `wicketd`.

Since this is a caching proxy for inventory, when wicket users request inventory a request gets forwarded over a channel to the `MgsManager` which returns an `Arc<RackV1Inventory>`. This prevents the need to copy this data. When the inventory gets updated a new `Arc<RackV1Inventory>` is created and will be served over the same channel. This channel based mechanism works nicely, but it's worth noting the alternative of using an `Arc<Mutex<` or `<Arc<RwLock`, or even a
`tokio::sync::watch::channel`. There may be performance tradeoffs with them, but the behavior is generally the same, and I find the message passing flow easier to follow in most cases. I'm open to changing this down the road.

There is also an intention to support ETAGs for the inventory, which would negate the need to bother with an `Arc` when sending the message over a channel, as copying would be so rare. This is left for a follow up commit, although there is some embedded support.

It should also be noted that polling the SPs every 10 seconds and retrieving inventory is really inefficient. Instead MGS should maintain a generation count of when inventory changes and provide information about what changed. Alternatively, MGS could also use ETAGs and just change the ETAG for `/sp` when anything changes. This would have more overhead, but *may* be simpler to implement. In order to implement either of these optimizations however, there would need to be some SP support for keeping track of changes. It's worth trying to determine when inventory can even change. Is it during upgrades or insertion of new hardware only?

This code was tested manually by running the following in separate terminals:
 * SP simulators (2 Sleds, 1 Switch)
 * MGS
 * wicketd
 * curl

```
➜  omicron git:(dfc01324) ✗ cargo run --bin sp-sim -- sp-sim/examples/config.toml
```

```
➜  omicron git:(dfc01324) ✗ cargo run --bin mgs -- run --id 14c8ad58-c55a-430c-9db5-5db2bfc3780e --address '[::1]:9999' gateway/examples/config.toml`
```

```
➜  omicron git:(dfc01324) ✗ cargo run -p wicketd -- run --address "[::1]:8000" wicketd/examples/config.toml
```

```
➜  ~ curl --silent "[::1]:8000/inventory" | jq
{
  "sps": [
    [
      {
        "details": {
          "state": "enabled",
          "serial_number": "00000000000000000000000000000002"
        },
        "info": {
          "details": {
            "present": "yes",
            "ctrl_detect_0": true,
            "ctrl_detect_1": false,
            "flt_a2": false,
            "flt_a3": false,
            "flt_rot": false,
            "flt_sp": false,
            "id": 17,
            "power": true
          },
          "id": {
            "slot": 0,
            "type": "sled"
          }
        }
      },
      []
    ],
    [
      {
        "details": {
          "state": "enabled",
          "serial_number": "00000000000000000000000000000001"
        },
        "info": {
          "details": {
            "present": "yes",
            "ctrl_detect_0": true,
            "ctrl_detect_1": false,
            "flt_a2": false,
            "flt_a3": false,
            "flt_rot": false,
            "flt_sp": false,
            "id": 18,
            "power": true
          },
          "id": {
            "slot": 0,
            "type": "switch"
          }
        }
      },
      []
    ],
    [
      {
        "details": {
          "state": "enabled",
          "serial_number": "00000000000000000000000000000003"
        },
        "info": {
          "details": {
            "present": "yes",
            "ctrl_detect_0": true,
            "ctrl_detect_1": false,
            "flt_a2": false,
            "flt_a3": false,
            "flt_rot": false,
            "flt_sp": false,
            "id": 17,
            "power": true
          },
          "id": {
            "slot": 1,
            "type": "sled"
          }
        }
      },
      []
    ]
  ]
}
```

### CODE REVIEW

File: gateway-client/src/lib.rs [hunk 1]
```diff
@@ -4,6 +4,47 @@
 
 // Copyright 2022 Oxide Computer Company
 
-//! Interface for API requests to a Management Gateway Service (MGS) instance
```
Comment (andrewjstone) [inline gateway-client/src/lib.rs:7]: This is an accident. I will fix it.

File: wicketd/src/mgs_manager.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+impl MgsManager {
+    pub fn new(log: &Logger, mgs_addr: SocketAddrV6) -> MgsManager {
+        let log = log.new(o!("component" => "wicketd MgsManager"));
+        let (tx, rx) = mpsc::channel(CHANNEL_CAPACITY);
+        let endpoint =
+            format!("http://[{}]:{}", mgs_addr.ip(), mgs_addr.port());
+        info!(log, "MGS Endpoint: {}", endpoint);
+        let mgs_client = gateway_client::Client::new(&endpoint, log.clone());
```
Comment (jgallagher) [inline wicketd/src/mgs_manager.rs:91]: Not suggesting a change here necessarily, but just noting from past experience - progenitor clients created via `::new()` implicitly have a [15 second timeout](https://github.com/oxidecomputer/progenitor/blob/4e2dcc5365ea4282af7b1851c5ae96305cb51d9a/progenitor-impl/src/lib.rs#L247-L253) on requests. If that's not what we want, we need to use `new_with_client()` to override it.

File: wicketd/src/mgs_manager.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    // We only want one outstanding inventory request at a time
+    let (tx, rx) = mpsc::channel(1);
+
+    tokio::spawn(async move {
+        let mut ticker = interval(MGS_POLL_INTERVAL);
+        ticker.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        loop {
+            ticker.tick().await;
+            match client.sp_list(MAX_COMPONENTS, None, Some(MGS_TIMEOUT)).await
```
Comment (jgallagher) [inline wicketd/src/mgs_manager.rs:141]: I know we already chatted about this a bit and are going to chat more, but just noting as I'm reading - this only fetches the first page of data.

**Review by jgallagher — APPROVED** [review]

**Comment by jgallagher** [comment]
> In order to implement either of these optimizations however, there would need to be some SP support for keeping track of changes. It's worth trying to determine when inventory can even change. Is it during upgrades or insertion of new hardware only?

We should talk more about this, but one question - what does "change" mean here? In the current implementation (and I _think_ this will remain true?), the list of components one gets back from an SP is fixed until there's an upgrade, and some of the properties of each component are similarly fixed (e.g., whether a given component is itself updateable) until there's an upgrade. But (a) the list of SPs responding to requests could change at any time, and (b) some of the properties of each component (like whether or not the component is present) could change at any time.

File: gateway-client/src/lib.rs [hunk 4]
```diff
@@ -4,6 +4,47 @@
...
+// case, we want to prohibit users from accidentally exposing implementation
+// specific, private types. As an example, we don't want to make it trivial
+// to include propolis types that may change in the Oxide public API. For
+// scenarios like this, forcing the user of a client API to generate a new type
+// serves as a safety feature in the common case. The user of an underlying
+// dropshot server can always choose to use this escape hatch.
+//
+//  In this case, we choose to allow the escape hatch, because one of two
+// primary consumers of MGS is wicketd, and wicketd wants to share inventory
+// data, which is already in a useful TLV format, directly with wicket for
```
[the remark this answers is not in the document]
Comment (andrewjstone) [reply gateway-client/src/lib.rs:24]: Yeah, you are right. I want to tighten up this comment anyway and will fix that.

File: wicketd/src/mgs_manager.rs (hunk 2 above)
Comment (andrewjstone) [reply wicketd/src/mgs_manager.rs:91]: Thanks for the heads up!

File: wicketd/src/mgs_manager.rs (hunk 3 above)
Comment (andrewjstone) [reply wicketd/src/mgs_manager.rs:141]: @jgallagher and I discussed this, and there's an open issue for converting this MGS endpoint to an unpaginated one. I'm going to do that next. https://github.com/oxidecomputer/omicron/issues/1913

**Comment by andrewjstone** [comment]
> > In order to implement either of these optimizations however, there would need to be some SP support for keeping track of changes. It's worth trying to determine when inventory can even change. Is it during upgrades or insertion of new hardware only?
> 
> We should talk more about this, but one question - what does "change" mean here? In the current implementation (and I _think_ this will remain true?), the list of components one gets back from an SP is fixed until there's an upgrade, and some of the properties of each component are similarly fixed (e.g., whether a given component is itself updateable) until there's an upgrade. But (a) the list of SPs responding to requests could change at any time, and (b) some of the properties of each component (like whether or not the component is present) could change at any time.

@jgallagher and I talked about this in detail. The inventory for devices reported by an SP comes from a TOML file in hubris. It has state that can change, but that should be infrequent. Right now the only field that can change is presence (which actually doesn't because it's hardcoded currently) and we don't particularly care about this being timely, especially for MUP. If a temperature sensor goes away, we can almost certainly get away with not knowing about it for 1 minute. In the future we should have some sort of alert system in a hubris task for things like this, but that's not necessary right now.

What we want to be more visible is something like someone pulling a sled. We can poll ignition more frequently for this and display it within a few seconds. We can poll the overall SPs slightly less frequently, or timeout quickly, since ignition is polled as part of the `/sp` endpoint.

**Comment by jgallagher** [comment]
> @jgallagher and I talked about this in detail. The inventory for devices reported by an SP comes from a TOML file in hubris. It has state that can change, but that should be infrequent. Right now the only field that can change is presence (which actually doesn't because it's hardcoded currently) and we don't particularly care about this being timely, especially for MUP. If a temperature sensor goes away, we can almost certainly get away with not knowing about it for 1 minute. In the future we should have some sort of alert system in a hubris task for things like this, but that's not necessary right now.

One nit that does not change any of the decisions/conclusions: the presence field is only hardcoded for a [few of the components](https://github.com/oxidecomputer/hubris/blob/master/task/control-plane-agent/src/inventory.rs#L149-L170); most of the components are getting live presence via the same path as `humility validate` so can change.

**Pull request merged**